src/dtls13.c

```diff
@@ -0,0 +1,1968 @@
+            if (ret != 0)
+                return ret;
+        }
+    }
+#endif
+
+    /* TODO: support chacha based ciphersuite */
+
+    return ret;
+}
+
+#define DTLS13_RN_EPOCH_SIZE (16 / 8)
+#define DTLS13_RN_SEQ_SIZE (48 / 8)
+
+static int Dtls13GetAckListLength(Dtls13RecordNumber *list, word16 *length)
+{
+    int numberElements;
+
+    numberElements = 0;
+
+    /* TODO: check that we don't exceed the maximum length */
+
+    while(list != NULL) {
+        list = list->next;
+        numberElements++;
+    }
+
+    *length = (DTLS13_RN_EPOCH_SIZE + DTLS13_RN_SEQ_SIZE) * numberElements;
+    return 0;
+}
+
+static int Dtls13WriteAckMessage(
+    WOLFSSL *ssl, Dtls13RecordNumber *recordNumberList, word32 *length)
+{
+    word16 msgSz, headerLength;
+    byte *output, *ackMessage;
+    word32 sendSz;
+    int ret;
+
+    sendSz = 0;
+
+    if (ssl->dtls13EncryptEpoch == NULL)
+        return BAD_STATE_E;
+
+    if (ssl->dtls13EncryptEpoch->epochNumber == 0) {
+        /* unprotected ACK */
+        headerLength = DTLS_RECORD_HEADER_SZ;;
+    }
+    else {
+        headerLength = Dtls13GetRlHeaderLength(ssl, 1);
+        sendSz += MAX_MSG_EXTRA;
+    }
+
+    ret = Dtls13GetAckListLength(recordNumberList, &msgSz);
+    if (ret != 0)
+        return ret;
+
+    sendSz += headerLength;
+
+    /* ACK list size */
+    sendSz += OPAQUE16_LEN;
+
+    /* ACK list */
+    sendSz += msgSz;
+
+    ret = CheckAvailableSize(ssl, sendSz);
+    if (ret != 0)
+        return ret;
+
+    output = ssl->buffers.outputBuffer.buffer +
+        ssl->buffers.outputBuffer.length;
+
+    ackMessage = output + headerLength;
+
+    c16toa(msgSz, ackMessage);
+    ackMessage += OPAQUE16_LEN;
+
+    while (recordNumberList != NULL) {
+        XMEMCPY(
+            ackMessage, recordNumberList->seq, sizeof(recordNumberList->seq));
+        ackMessage += sizeof(recordNumberList->seq);
+        recordNumberList = recordNumberList->next;
+    }
+
+    *length = msgSz + OPAQUE16_LEN;
+
+    return 0;
+}
+
+static int Dtls13RtxIsTrackedByRn(Dtls13RtxRecord *r, const word32 seq[2])
+{
+    int i;
+
+    for (i = 0; i < r->rnIdx; ++i) {
+        if (r->rn[i * 2] == seq[0] && r->rn[i * 2 + 1] == seq[1]) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+static int Dtls13RtxRemoveRecord(
+    WOLFSSL *ssl, Dtls13RtxFSM *fsm, const word32 seq[2])
+{
+    Dtls13RtxRecord *head, *r;
+
+    head = fsm->rtxRecords;
+
+    if (head == NULL)
+        return 0;
+
+    /* check head first */
+    if (Dtls13RtxIsTrackedByRn(head, seq)) {
+        fsm->rtxRecords = head->next;
+        Dtls13FreeRtxBufferRecord(ssl, head);
+        return 0;
+    }
+
+    while (head->next != NULL) {
+        r = head->next;
+        if (Dtls13RtxIsTrackedByRn(r, seq)) {
+            head->next = r->next;
+            Dtls13FreeRtxBufferRecord(ssl, r);
+            return 0;
+        }
+
+        head = head->next;
+    }
```

Comment:
Let's make this one `while` loop
```
while (head != NULL) {
  ...
  head = head->next;
}
```

Reply:
If I do, I should keep track of the preceding element to  do the unlink `head->next = r->next;`